### PR TITLE
[4.0] A more flexible ActionButton class

### PR DESCRIPTION
**Pull request by asika32764**
## Introduction

Currently we use `JHtml` (`HTMLHelper`) to make the action buttons in grid view. 

p-2017-11-05-001

But the interface and the number of arguments of `JHtmlJGrid::published()` is horrible, it hard to use and extends, developers must read many code to understand how to make their own buttons.

## The Solution: `ActionButton`

I try to create a class that is configurable, this is a sample code:

```php
<?php
// In View class
$button = (new ActionButton(['disabled' => !$canChange]))
    ->addState(0, 'articles.featured', 'icon-unfeatured', 'COM_CONTENT_UNFEATURED');
    ->addState(1, 'articles.unfeatured', 'icon-featured', 'COM_CONTENT_FEATURED');

$this->featuredButton = $button;
```

```php
// In template table view loop

<?php foreach ($articles as $i => $article): ?>
    ...
   <td><?php echo $this->featuredButton->render($article->featured, $i); ?></td>
    ...
<?php endforeach; ?>
```

It's simple and easy to learn, make developers days joy.

## Layout

This class use Joomla layouts to render template. Also we can override layout:

```php
$button->setLayout('my.action.button');
```

## Default State

If an item has an unexpected state value, for example, I changed an article state to `5`, it will show `Unknow State` icon:

p-2017-11-05-002

## `PublishedButton`

For standard Joomla content, I also created a `PublishedButton` that can handle the publish time:

```php
$button = new PublishedButton(
	[
		'diabled' => $canDoWhat,
		'task_prefix' => 'articles.',
		'checkbox_name' => 'cb'
	]
);

echo $button->render($value, $i, $publish_up, $publish_down);
```

p-2017-11-23-003

-------

In this PR, I rewrite articles `published` and `featured` button to use this class as example, it works well.

If this concept accepted, I will continue finish the test and docblock.

### CODE REVIEW

File: libraries/vendor/joomla/utilities/src/ArrayHelper.php [hunk 1]
```diff
@@ -744,4 +744,42 @@ public static function flatten($array, $separator = '.', $prefix = '')
...
+	 *
+	 * @param   array  ...$args  Array to be merge.
+	 *
+	 * @return  array  Merged array.
+	 *
+	 * @throws \InvalidArgumentException
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public static function mergeRecursive(...$args)
```
Comment (asika32764) [inline libraries/vendor/joomla/utilities/src/ArrayHelper.php]: I know this is framework file, so I will create PR to framework package if this PR accepted.

**Comment by dgrammatiko** [comment]
Hmm, I have something different in my mind for j4: using custom elements and client side rendering...

**Comment by asika32764** [comment]
I think it won't be conflict. We still must call PHP to do some configuration or translate string before we call any client-side functions. Just change the template to raise any HTML/JS handling. This class is just a PHP entry to help developers enable button actions.

**Comment by asika32764** [comment]
So do you have any plan for the `JToolbar` ? I also have some idea similar to this concept.

```php
$toolbar = JToolbar::getInstance('administrator')
    ->addButton(new AddNewButton)
    ->addButton(new PublishButton)
    ->addButton(new CustomButton($html, JText::_('XXX')));

$toolbar->render();
```

**Comment by dgrammatiko** [comment]
I've started some work like 3 months ago: https://github.com/joomla/joomla-cms/compare/4.0-dev...dgt41:§4.0-dev-toolbar-cleanup?expand=1

Actually right now I think that the toolbar should be utilising custom elements, which obviously the above link is not...

**Comment by asika32764** [comment]
I take a look your PR, I think my code can get together with custom elements without any problem.

See https://github.com/joomla/joomla-cms/pull/17375/files#diff-cffc36facfda8c2f198b4f893dd21bbaR35

Custom elements still needs PHP code to render it, most of all, we use `JHtml` to render HTML elements and include JS files. The main subject or this PR is to replace JHtml and make API interface configurable.

We can imagine the button layout integrated the custom element will look like:

```php
<?php

JHtml::_('webcomponent', ['toolbar-button' => 'system/joomla-toolbar-button.min.js'], ['relative' => true, 'version' => 'auto']);

extract($displayData);
?>
<toolbar-button
  id="<?php echo $id;?>"
  class="<?php echo $class;?>"
  onclick="<?php echo $click;?>"
  color="success"
  check-list="true"
  check-list-warning="Please select an item.">
<?php echo $text;?>
</toolbar-button>
```

**Comment by wilsonge** [comment]
I like this :) It seems a lot more obvious than what we had before!

File: libraries/vendor/joomla/utilities/src/ArrayHelper.php (hunk 1 above)
Comment (wilsonge) [reply libraries/vendor/joomla/utilities/src/ArrayHelper.php]: Can we do the framework PR please :) I'd like to merge this PR into core!

File: libraries/cms/html/jgrid.php [hunk 2]
```diff
@@ -194,7 +196,7 @@ public static function published($value, $i, $prefix = '', $enabled = true, $che
 				$tips[] = JText::sprintf('JLIB_HTML_PUBLISHED_FINISHED', JHtml::_('date', $publish_down, JText::_('DATE_FORMAT_LC5'), 'UTC'));
 			}
 
-			$tip = empty($tips) ? false : implode('<br>', $tips);
+			$tip = empty($tips) ? false : implode('<br />', $tips);
```
Comment (Quy) [inline libraries/cms/html/jgrid.php]: @brianteeman please confirm to use `<br>` and not `<br />` in v4.0.

Comment (brianteeman) [reply libraries/cms/html/jgrid.php]: yes it is without the slash

Comment (asika32764) [reply libraries/cms/html/jgrid.php]: Hmm, actually I didn't touch this. I think it is merged from a current branches.

File: libraries/vendor/joomla/utilities/src/ArrayHelper.php (hunk 1 above)
Comment (asika32764) [reply libraries/vendor/joomla/utilities/src/ArrayHelper.php]: Already did and merged here:
https://github.com/joomla-framework/utilities/pull/20

File: administrator/components/com_content/tmpl/articles/default.php [hunk 3]
```diff
@@ -49,6 +52,14 @@
 }
 
 $assoc = JLanguageAssociations::isEnabled();
+
+// Configure content stet renderer.
```
Comment (Quy) [inline administrator/components/com_content/tmpl/articles/default.php]: `stet` maybe `state`?

File: libraries/src/Button/ActionButton.php [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 *
+	 * @throws  \InvalidArgumentException
+	 */
+	public function render(string $value = '', string $row = null, array $options = []): string
+	{
+		$data = $this->getState($value);
+
+		$data = $data ? : $this->getState('_default');
```
Comment (Quy) [inline libraries/src/Button/ActionButton.php]: change to `?:`

File: libraries/src/Button/ActionButton.php [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+	 */
+	public function setLayout(string $layout): self
+	{
+		$this->layout = $layout;
+
+		return $this;
+	}
+
+	/**
+	 * Method to get property Options
```
Comment (Quy) [inline libraries/src/Button/ActionButton.php]: lowercase `Options`

File: libraries/src/Button/ActionButton.php (hunk 4 above)
Comment (asika32764) [reply libraries/src/Button/ActionButton.php]: So Joomla changed code style to `?:` ???

Comment (Quy) [reply libraries/src/Button/ActionButton.php]: I did a searched on `? :` and found a handful of them, However, the majority are `?:`.

Comment (dgrammatiko) [reply libraries/src/Button/ActionButton.php]: J4 is PHP 7+ so we can use it here

Comment (mbabker) [reply libraries/src/Button/ActionButton.php]: Yes use `?:`.  We already have a verbose enough code style with a hard line limit that forces creative concatenation of things, any place we can save a character (or 5 in the case of 4.x with short array syntax) is a good thing.

**Comment by wilsonge** [comment]
Thankyou very much for this :)

**Pull request merged**